examples/node2vec_link.py

```diff
@@ -0,0 +1,154 @@
+parser.add_argument("--embedding_dim", type=int, default=128)
+parser.add_argument("--walk_length", type=int, default=10)
+parser.add_argument("--context_size", type=int, default=5)
+parser.add_argument("--num_workers", type=int, default=1)
+parser.add_argument("--log_dir", type=str, default="results")
+args = parser.parse_args()
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+if torch.cuda.is_available():
+    torch.set_num_threads(1)
+seed_everything(42)
+
+root_dir = "./data"
+
+# TODO: remove process=True once correct data/task is uploaded.
+dataset: RelBenchDataset = get_dataset(name=args.dataset, process=True)
+task: LinkTask = dataset.get_task(args.task, process=True)
+tune_metric = "link_prediction_map"
+assert task.task_type == TaskType.LINK_PREDICTION
+
+col_to_stype_dict = dataset2inferred_stypes[args.dataset]
+
+# We don't need to specify any text column or use any text
+# embeddings as features are not used in node2vec baseline
+for table_name, table_col_to_stype in col_to_stype_dict.items():
+    for col, stype in table_col_to_stype.items():
+        if stype == torch_frame.text_embedded:
+            col_to_stype_dict[table_name][col] = torch_frame.categorical
+
+
+data, col_stats_dict = make_pkey_fkey_graph(
+    dataset.db,
+    col_to_stype_dict=col_to_stype_dict,
+    cache_dir=os.path.join(root_dir, f"{args.dataset}_materialized_cache"),
+)
+
+
+num_src_nodes = task.num_src_nodes
+df = task.train_table.df.explode(task.dst_entity_col)
+# edge from src to dst
+edge_index = torch.stack(
+    [
+        torch.from_numpy(df[task.src_entity_col].astype(int).values),
+        (torch.from_numpy(df[task.dst_entity_col].astype(int).values + num_src_nodes)),
+    ]
+)
+
+
+model = Node2Vec(
+    edge_index=to_undirected(edge_index),
+    embedding_dim=args.embedding_dim,
+    walk_length=args.walk_length,
+    context_size=args.context_size,
+    num_negative_samples=1,
+    p=1.0,
+    q=1.0,
+    sparse=True,
+).to(device)
+
+loader = model.loader(
+    batch_size=args.batch_size, shuffle=True, num_workers=args.num_workers
+)
+optimizer = torch.optim.SparseAdam(model.parameters(), lr=args.lr)
+
+
+def train():
+    model.train()
+    total_loss = 0
+    for pos_rw, neg_rw in loader:
+        optimizer.zero_grad()
+        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
+        loss.backward()
+        optimizer.step()
+        total_loss += loss.item()
+    return total_loss / len(loader)
+
+
+def get_valid_test_dst_entities():
+    dst_entity_table = task.dataset.db.table_dict[task.dst_entity_table]
+    # Might be good to filter out entities based on time col in the future
+    return dst_entity_table.df[task.dst_entity_col].values + num_src_nodes
+
+
+@torch.no_grad()
+def test(table: Table, k):
+    model.eval()
+    z = model()
+    src = z[table.df[task.src_entity_col].values]
+    dst = z[get_valid_test_dst_entities()]
+    _, indices = torch.topk(src @ dst.T, k)
```

Comment:
Node2vec is optimized to predict random walk probability. It's not explicitly optimizing inner-product-based prediction accuracy. It's a bit odd to directly use node2vec's embeddings to make prediction.

Instead, what people usually do is to pre-train node2vec embeddings and then train another classifier on top of the node2vec embeddings to make the final prediction. @rusty1s can give more details on implementation.